apollo-router/src/query_planner/plan.rs

```diff
@@ -468,7 +468,6 @@ impl PlanNode {
         Ok(())
     }
 
-    #[cfg(test)]
     /// Retrieves all the services used across all plan nodes.
```

Comment:
This was just test code before - are we comfortable making that production code? It does a lot of recursion and heap allocation, which could be slow on very large query plans.

Reply:
Thanks for the review! I’m realizing now that we should probably do this differently:

* having it on the context means it’s always created even if unused, and will be impossible to remove
* it should probably be a field on the coprocessor request body so that it’s opt-in and lazily evaluated
* It should probably be generic for subgraphs and subgraph/sources, so that you can fetch auth tokens for any fetch target 

I’ll refactor, and I’ll take a closer look at this function to see if there’s a more efficient way to get the service names from the plan

Reply:
after discussion in slack, i went back to the context-based approach but added a config option to opt into this behavior. (i also implemented a non-recursive version of this method that allocates only a set and a vec.)